GenderPayGap.WebUI/BackgroundJobs/ScheduledJobs/AnonymiseThreeYearOldFeedbackJob.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using GenderPayGap.Core.Interfaces;
+using GenderPayGap.Database.Models;
+using GenderPayGap.Extensions;
+
+namespace GenderPayGap.WebUI.BackgroundJobs.ScheduledJobs
+{
+    public class AnonymiseThreeYearOldFeedbackJob
+    {
+
+        private readonly IDataRepository dataRepository;
+
+        public AnonymiseThreeYearOldFeedbackJob(IDataRepository dataRepository)
+        {
+            this.dataRepository = dataRepository;
+        }
+
+        public void AnonymiseFeedbackAction()
+        {
+            var runId = JobHelpers.CreateRunId();
+            var startTime = VirtualDateTime.Now;
+            JobHelpers.LogFunctionStart(runId, nameof(AnonymiseFeedbackAction), startTime);
+
+            try
+            {
+                DateTime threeYearsAgo = VirtualDateTime.Now.AddYears(-3);
+
+                List<Feedback> feedback = dataRepository.GetAll<Feedback>()
+                    .Where(f => DateTime.Compare(f.CreatedDate, threeYearsAgo) <= 0)
+                    .Where(f => !f.HasBeenAnonymised)
+                    .ToList();
+
+                foreach (Feedback feedbackItem in feedback)
+                {
+                    AnonymiseFeedback(feedbackItem);
+                }
+
+                JobHelpers.LogFunctionEnd(runId, nameof(AnonymiseFeedbackAction), startTime);
+            }
+            catch (Exception ex)
+            {
+                JobHelpers.LogFunctionError(runId, nameof(AnonymiseFeedbackAction), startTime, ex);
+
+                //Rethrow the error
+                throw;
+            }
+        }
+
+        public void AnonymiseFeedback(Feedback feedback)
+        {
+            feedback.OtherSourceText = string.IsNullOrWhiteSpace(feedback.OtherSourceText)
+                ? "not supplied"
+                : "supplied";
+
+            feedback.OtherReasonText = string.IsNullOrWhiteSpace(feedback.OtherReasonText)
+                ? "not supplied"
+                : "supplied";
+
+            feedback.OtherPersonText = string.IsNullOrWhiteSpace(feedback.OtherPersonText)
+                ? "not supplied"
+                : "supplied";
+
+            feedback.EmailAddress = string.IsNullOrWhiteSpace(feedback.EmailAddress)
+                ? "not supplied"
+                : "supplied";
+
+            feedback.PhoneNumber = string.IsNullOrWhiteSpace(feedback.PhoneNumber)
+                ? "not supplied"
+                : "supplied";
+
+            feedback.Details = string.IsNullOrWhiteSpace(feedback.Details)
+                ? "not supplied"
+                : "supplied";
+
+            feedback.HasBeenAnonymised = true;
+
+            dataRepository.SaveChangesAsync().Wait();

```

Comment:
It will be quicker to do the Save outside of the loop (i.e. in the parent method) - that way, we save everything in one go.

There's a slightly irritating thing here about Entity Framework.
EF is very clever and tracks changes to your objects, so it knows what to save.
But, it doesn't know which objects you've finished dealing with.
* The 1st time round this loop, we'd update Feedback 1 and save it - EF would check Feedback 1 and save it
* The 2nd time round this loop, we'd update Feedback 1 and save it - EF would check Feedback 1 and Feedback 2 and save just 2 (but it would have to check both Feedback 1 and 2)

After a while, EF ends up checking loads of objects each time, which gets quite inefficient.
So it's often better to do just one big save at the end.